### ISSUE TITLE
followup: BS & prototype FP

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [arkenfox/TZP#35]
as discussed elsewhere
* prototype fingerprint
   - [x] add it
   - [x] hash it, linkify counts e.g. `x lies` from `y properties`, `details`
   - [x] fixup Opera false positives
   - [ ] apply it to metrics
   - [ ] add a `[click here]` manually test for the full check
* same but for mathematical and other BS PoCs
   - [x] stash in a global array, output somewhere: not needed for a fingerprint since these are already baked into the results
   - [x] pluginBS + mimeBS
   - [x] each canvas / isPoint method
   - [x] uaBS
   - [x] each domrect method: shift
   - [x] each domrect method: known value (FF only: unless abraham has cracked it)
   - [x] navigator fakeKeys (thanks abraham)
   - [ ] font lies : note `perspective` is a false positive in FF61- (from memory)
   - [ ] each textmetric method
   * various language section lies
      - [ ] first I want to reduce the metrics here for the FP to: timezone, maybe epoch hash, language, additional languages, locale, geo properties: everything else is used to detect lies (but most of the tests are a mix: to check RFP is working)
      - [ ] and I need to add an iframe test
      - [ ] timezoneBS
      - [ ] languageBS / localeBS - use Intl.DisplayNames to compare?
      - [ ] additional languages would be hard to detect

just on the BS stuff: there's more to come: textmetrics lies per method, and I have some cool ideas for language PoCs
- there's also an audio lie PoC I have in mind

**Comment by Thorin-Oakenpants** [comment]
@abrahamjuliot can you explain the diff between your `lies` and `trash`

nvm

> Trash
>     unusual results
>     forgivable lies
> 
> Lies
>     prototype tampering
>     failed math calculations

OK, so this is basically the same thing. I can collect failed math/PoCs (domrect shift, timezone PoC, known canvas results PoC, etc) in one array and we can build a comprehensive prototype PoC for another array
- each can be output as an actual FP item within their own right

**Comment by abrahamjuliot** [comment]
Here's the `getLies` function rebuilt from the ground up. It searches everything.

https://jsfiddle.net/zcw45nfu/

Use in this fashion
```js
const { lieList, lieDetail } = getLies() // execute and destructure the list and detail

// log to see the goods and analyze
console.log(lieList)
console.log(lieDetail)

// check lies later in any function
lieList.includes('HTMLCanvasElement.toDataURL') // returns true or false
lieDetail['HTMLCanvasElement.toDataURL'] // returns the list of lies
```

<details close>
<summary>View the code</summary>

```js
const getLies = () => {
    // Lie Tests

    // object constructor descriptor should return undefined properties
    const getUndefinedValueLie = (obj, name) => {
        const objName = obj.name
        const objNameUncapitalized = window[objName.charAt(0).toLowerCase() + objName.slice(1)]
        const hasInvalidValue = !!objNameUncapitalized && (
            typeof Object.getOwnPropertyDescriptor(objNameUncapitalized, name) != 'undefined' ||
            typeof Reflect.getOwnPropertyDescriptor(objNameUncapitalized, name) != 'undefined'
        )
        return hasInvalidValue ? true : false
    }

    // creating a new instance of the API function should throw a TypeError
    const getNewInstanceTypeErrorLie = apiFunction => {
        let newTypeError
        try {
            new apiFunction()
            return true
        } catch (error) {
            return error.constructor.name != 'TypeError' ? true : false
        }
    }

    // extending the API function to a fake class should throw a TypeError
    const getClassExtendsTypeErrorLie = apiFunction => {
        try {
            class Fake extends apiFunction {}
            return true
        } catch (error) {
            // Native has TypeError and 'not a constructor' message in FF & Chrome
            return error.constructor.name != 'TypeError' ? true :
                !/not a constructor/i.test(error.message) ? true : false
        }
    }

    // setting prototype to null and converting to primitive value should throw a TypeError
    const getNullConversionTypeErrorLie = apiFunction => {
        const nativeProto = Object.getPrototypeOf(apiFunction)
        try {
            Object.setPrototypeOf(apiFunction, null) + ''
            return true
        } catch (error) {
            return error.constructor.name != 'TypeError' ? true : false
        } finally {
            // restore proto
            Object.setPrototypeOf(apiFunction, nativeProto)
        }
    }

    // toString() and toString().toString() should return a native string
    const getToStringLie = (apiFunction, name) => {
        /*
        Accepted strings:
        'function name() { [native code] }'
        'function name() {\n    [native code]\n}'
        'function get name() { [native code] }'
        'function get name() {\n    [native code]\n}'
        'function () { [native code] }'
        `function () {\n    [native code]\n}`
        */
        const nativeString = new RegExp(
            `^function (get ${name}|${name}|)\\(\\) {(\\n    | )\\[native code\\](\\n| )}$`
        )
        return (
            !nativeString.test(apiFunction.toString()) ||
            !nativeString.test(apiFunction.toString().toString())
        )
    }

    // 'prototype' should not exist in API function
    const getPrototypeInFunctionLie = apiFunction => 'prototype' in apiFunction ? true : false

    // 'arguments', 'caller', 'prototype', 'toString' should not exist in descriptor
    const getDescriptorLie = apiFunction => {
        const hasInvalidDescriptor = (
            !!Object.getOwnPropertyDescriptor(apiFunction, 'arguments') ||
            !!Reflect.getOwnPropertyDescriptor(apiFunction, 'arguments') ||
            !!Object.getOwnPropertyDescriptor(apiFunction, 'caller') ||
            !!Reflect.getOwnPropertyDescriptor(apiFunction, 'caller') ||
            !!Object.getOwnPropertyDescriptor(apiFunction, 'prototype') ||
            !!Reflect.getOwnPropertyDescriptor(apiFunction, 'prototype') ||
            !!Object.getOwnPropertyDescriptor(apiFunction, 'toString') ||
            !!Reflect.getOwnPropertyDescriptor(apiFunction, 'toString')
        )
        return hasInvalidDescriptor ? true : false
    }

    // 'arguments', 'caller', 'prototype', 'toString' should not exist as own property
    const getOwnPropertyLie = apiFunction => {
        const hasInvalidOwnProperty = (
            apiFunction.hasOwnProperty('arguments') ||
            apiFunction.hasOwnProperty('caller') ||
            apiFunction.hasOwnProperty('prototype') ||
            apiFunction.hasOwnProperty('toString')
        )
        return hasInvalidOwnProperty ? true : false
    }

    // descriptor keys should only contain 'length' and 'name' 
    const getDescriptorKeysLie = apiFunction => {
        const descriptorKeys = Object.keys(Object.getOwnPropertyDescriptors(apiFunction))
        const hasInvalidKeys = '' + descriptorKeys != 'length,name' && '' + descriptorKeys != 'name,length'
        return hasInvalidKeys ? true : false
    }

    // own property names should only contain 'length' and 'name' 
    const getOwnPropertyNamesLie = apiFunction => {
        const ownPropertyNames = Object.getOwnPropertyNames(apiFunction)
        const hasInvalidNames = (
            '' + ownPropertyNames != 'length,name' && '' + ownPropertyNames != 'name,length'
        )
        return hasInvalidNames ? true : false
    }

    // own keys names should only contain 'length' and 'name' 
    const getOwnKeysLie = apiFunction => {
        const ownKeys = Reflect.ownKeys(apiFunction)
        const hasInvalidKeys = '' + ownKeys != 'length,name' && '' + ownKeys != 'name,length'
        return hasInvalidKeys ? true : false
    }

    // API Function Test
    const getPrototypeLies = (apiFunction, obj = null) => {
        if (typeof apiFunction != 'function') {
            return {
                lied: false,
                lieTypes: []
            }
        }
        const name = apiFunction.name.replace(/get\s/, '')
        const lies = {
            // custom lie string names: adjust if desired
            'failed undefined value': obj ? getUndefinedValueLie(obj, name) : false,
            'failed new instance type error': getNewInstanceTypeErrorLie(apiFunction),
            'failed class extends type error': getClassExtendsTypeErrorLie(apiFunction),
            'failed null conversion type error': getNullConversionTypeErrorLie(apiFunction),
            'failed to string': getToStringLie(apiFunction, name),
            'failed prototype in function': getPrototypeInFunctionLie(apiFunction),
            'failed descriptor': getDescriptorLie(apiFunction),
            'failed own property': getOwnPropertyLie(apiFunction),
            'failed descriptor keys': getDescriptorKeysLie(apiFunction),
            'failed own property names': getOwnPropertyNamesLie(apiFunction),
            'failed own keys': getOwnKeysLie(apiFunction)
        }
        const lieTypes = Object.keys(lies).filter(key => !!lies[key])
        return {
            lied: lieTypes.length,
            lieTypes
        }
    }

    // Lie Detector
    const createLieDetector = () => {
        const props = {} // lie list and detail
        return {
            getProps: () => props,
            searchLies: (obj, {
                ignore
            } = {}) => Object.getOwnPropertyNames(obj.prototype).forEach(name => {
                if (name == 'constructor' || (ignore && new Set(ignore).has(name))) {
                    return
                }
                const apiName = `${obj.name}.${name}`
                try {
                    const proto = obj.prototype
                    let res // response from getPrototypeLies

                    // search if function
                    try {
                        const apiFunction = proto[name] // may trigger TypeError
                        if (typeof apiFunction == 'function') {
                            res = getPrototypeLies(proto[name])
                            if (res.lied) {
                                return (props[`${obj.name}.${name}`] = res.lieTypes)
                            }
                            return
                        }
                    } catch (error) {}
                    // else search getter function
                    const getterFunction = Object.getOwnPropertyDescriptor(proto, name).get
                    res = getPrototypeLies(getterFunction, obj) // send the obj for special tests
                    if (res.lied) {
                        return (props[apiName] = res.lieTypes)
                    }
                    return
                } catch (error) {
                    // API may be blocked or unsupported
                    return console.log(apiName)
                }
            })

        }
    }

    const lieDetector = createLieDetector()
    const {
        searchLies
    } = lieDetector

    // search for lies: add properties to ignore if desired
    searchLies(Node)
    searchLies(Element)
    searchLies(HTMLElement)
    searchLies(HTMLCanvasElement)
    searchLies(Navigator)
    searchLies(Screen)
    searchLies(Date)
    searchLies(Intl.DateTimeFormat)
    searchLies(Intl.RelativeTimeFormat)
    searchLies(CanvasRenderingContext2D)
    searchLies(PluginArray)
    searchLies(Plugin)
    searchLies(Document)

    // if supported
    if ('WebGLRenderingContext' in window) {
        searchLies(WebGLRenderingContext)
    }
    if ('WebGL2RenderingContext' in window) {
        searchLies(WebGL2RenderingContext)
    }
    if ('OffscreenCanvasRenderingContext2D' in window) {
        searchLies(OffscreenCanvasRenderingContext2D)
    }
    if ('AnalyserNode' in window) {
        searchLies(AnalyserNode)
    }
    if ('AudioBuffer' in window) {
        searchLies(AudioBuffer)
    }

    // return lies list and detail 
    const props = lieDetector.getProps()
    return {
        lieList: Object.keys(props),
        lieDetail: props
    }
}

// start program
const {
    lieList,
    lieDetail
} = getLies()

console.log(lieList) // log the list of APIs tampered with
console.log(lieDetail) // view the lie type detail

// check lies
lieList.includes('HTMLCanvasElement.toDataURL') // returns true or false
lieDetail['HTMLCanvasElement.toDataURL'] // returns the list of lies

```
</details>

More APIs can be added to the `searchLies` section like so:
```js
searchLies(MediaDevices)
searchLies(RTCPeerConnection, {
    ignore: [
        // optionally ignore and skip a set of properties
        'pendingRemoteDescription'
        'pendingLocalDescription'
    ]
})
```
<details close>
<summary>See some result</summary>
<br>
CB

![image](https://user-images.githubusercontent.com/6946045/104550426-acd03800-55e9-11eb-935f-c1d6274341a6.png)


Chameleon

![image](https://user-images.githubusercontent.com/6946045/104553855-3551d700-55f0-11eb-8c96-5018f1cadcd5.png)

Trace

![image](https://user-images.githubusercontent.com/6946045/104553729-06d3fc00-55f0-11eb-9312-ff37f9bdb81d.png)

Cydec

![image](https://user-images.githubusercontent.com/6946045/104553938-56b2c300-55f0-11eb-91fc-5dc6f2104d55.png)

Privacy Possom

![image](https://user-images.githubusercontent.com/6946045/104554064-82ce4400-55f0-11eb-939e-464f488d9ef5.png)

DuckDuckGo Privacy Essentials

![image](https://user-images.githubusercontent.com/6946045/104554145-ab563e00-55f0-11eb-9411-8ff22a5e290e.png)

Tor Browser

![image](https://user-images.githubusercontent.com/6946045/104554388-f8d2ab00-55f0-11eb-84db-970471984f4b.png)

</details>

**Comment by Thorin-Oakenpants** [comment]
Tor Browser

Uncaught (in promise) ReferenceError: MediaDevices is not defined
    getPrototypeLies https://abrahamjuliot.github.io/creepjs/tests/prototype.js:297
    <anonymous> https://abrahamjuliot.github.io/creepjs/tests/prototype.js:340
    <anonymous> https://abrahamjuliot.github.io/creepjs/tests/prototype.js:387

**Comment by abrahamjuliot** [comment]
updated the fiddle and code above:
- moved `MediaDevices` to "if supported" section
- added search computation support for non-function objects like `Math` 
- added `iframeWindow` to aggressively detect failed `toString` tampering (per madtech's [technique](https://adtechmadness.wordpress.com/2019/03/23/javascript-tampering-detection-and-stealth/))
- handles null constructors (an insensitive form of tampering)

**Comment by Thorin-Oakenpants** [comment]
I've just been using your test page: am keen to add it to TZP as a promise before anything else kicks off: I guess that'll add 80ms or so

nightly: 0 lies detected in 1934 properties
my FF84: 0 lies detected in 1117 properties

wooo .. that's a lot of less properties

**Comment by Thorin-Oakenpants** [comment]
OK, so do we need to sort the results before hashing?

**Comment by abrahamjuliot** [comment]
As far as I know, sorting is not beneficial here. Each object interface should have a stable order and structure (unlike the window object properties). I'll add some hashes to the test page.

**Comment by Thorin-Oakenpants** [comment]
nightly now at 0 lies detected in 2398 properties

**Comment by Thorin-Oakenpants** [comment]
https://abrahamjuliot.github.io/creepjs/tests/prototype.html

wha the? ... 0 lies detected in 171 properties (FF nightly)

**Comment by Thorin-Oakenpants** [comment]
^ so how come one says 171 properties and the other says 164? AFAICT chameleon is not set to block anything
^ how come we went from a billion gazillion properties to a handful?

**Comment by abrahamjuliot** [comment]
I refactored the search code above:
- added target option to improve performance (removed ignore option)
  - I created a shortlist of targets based on results from a handful of extensions
  - to search all properties in an interface, leave out the option argument
  - some interfaces have only a dozen or a few dozen properties, so the target option is not necessary
  - there is an excessive number of properties in `Document`, `Node`, `Element`, `HTMLElement`, webgl/canvas `...RenderingContext...` (I recommend a target on these at minimum)
  - check the length of an interface with `Object.getOwnPropertyNames(interfaceObject).length`
- moved support/blocking checks to the search function: no longer need to figure out what needs support checking

```js
// search all
searchLies(() => Document) // search all ~200 properties

// target based approach: with the target option
searchLies(() => Document, {
        target: [
            'createElement',
            'createElementNS',
            'getElementById',
            'getElementsByClassName',
            'getElementsByName',
            'getElementsByTagName',
            'getElementsByTagNameNS',
            'referrer',
            'write',
            'writeln'
        ]
    })
```

**Comment by abrahamjuliot** [comment]
It appears properties are removed in Chameleon.

[diffchecker](https://www.diffchecker.com/XDrm8WXu) (with few settings on)

![image](https://user-images.githubusercontent.com/6946045/105124933-0aec9780-5a90-11eb-839c-a94e9353fa12.png)

**Comment by Thorin-Oakenpants** [comment]
I'm just waiting for you to perfect it all 😁 and then I can swoop in and add it like the hero I am

**Comment by Thorin-Oakenpants** [comment]
I can't wait TBH ... for everything it's like run test, apply mathematical proof where applicable (uaBS, pluginBS, mimeBS, etc), and then if noLies for that metric, check the prototypeLies

BTW: textmetrics are lied about in Brave. I have my PoC which I need to refine, but ATM the brave one just flat out makes the results all the same, except `width` which is also spoofed but different. The textmetric results are all integers (similar to what linux always returns). This may defeat my mathematical PoC, which needs some serious work to weed out false positives

PS: TZP textmetrics does not current have my math PoC: and maybe we could talk about it elsewhere to refine it: but basically the width of a single char should be double the width of the char repeated twice, and nothing should be more than 13 decimal places (or it might be 14, I forget). In the four `actualBounding` measurements, results should be cleanly divisible in `1/128ths`

However, depending on the default font (or we could use and check Aria or Times New Roman is used on desktop, since that seems universal - still leaves android hanging a bit, and of course what if document fonts are disabled), some chars cause false positives

Was trying to build my PoC to
- cycle thru fonts (and I'm limited to what is on the system)
   - cycle thru styles (serif, etc)
     - cycle thru chars (no weird shit like arabic or hebrew or joining chars etc)
- record
   - if single char width = `char+char` width, and check number of decimals
   - check for number of decimals in actualBounding
   - check 1/128ths
   - keep track of what is an integer (cuz they become useless)
- on each font loop, remove false positives from the char set

But that's doing my head in :)  maybe I should open this elsewhere

**Comment by abrahamjuliot** [comment]
I added some final touches above
- new illegal TypeError test "accessing the property from the prototype should throw a TypeError"
  - example: since the property is not a function, `Navigator.prototype.userAgent` should throw an error
- since blocked and unsupport is checked up front, an execution error now translates to a tampering lie:
  - "z: prototype tests should not fail execution"

**Comment by Thorin-Oakenpants** [comment]
Sweet. So...
- add `prototype.js`
- add it to the js files count so we know when to kick things off
- run it as a promise (which fills in a global map?)
   - how do I run it: didn't you say there were two modes? I am confused
- fingerprinting section: output the hash (div already there) with clickable `[x lies]` link for details
   - your details looks ok, but then gets repetitive with the `f,g,h,i` etc, but I can worry about this later
- and only after it has completed (promise), let the rest of the page load since we will want to use it for results

you show three hashes: a single hash should cover it for entropy, right? And you just output three for detail/research purposes
```
308 lies detected in 45 of 164 properties

164 searched: 8edce4b5
45 corrupted: 5714979a
308 lies: 7095ac7c
```

Note: the clickable `[x lies]` and `[x metrics]` links currently output the appropriate array (a name is passed as a parameter) to console, but when I work out how to do the css (css hurts my head), then it will output to a re-usable modal overlay with a scrollable inner and buttons to `copy to clipboard` and `log to console`, as well as a view-type: e.g. JSON, array, etc)

Still confused over dropping from ~2400 to way less properties checked. I get the perf aspect, but what isn't covered in the small one? Maybe I can add a link for the deep protoypeLies test. Or am I missing something here?

**Comment by Thorin-Oakenpants** [comment]
I am correct in assuming we need this to finish before we do anything else, right?

```
// check lies later in any function
lieList.includes('HTMLCanvasElement.toDataURL') // returns true or false
lieDetail['HTMLCanvasElement.toDataURL'] // returns the list of lies
```

**Comment by abrahamjuliot** [comment]
> ...finish before we do anything else, right?

Correct. This runs before all other tests. 

> run it as a promise 

Optional, but a promise version or an internal Promise.all on `searchLies` has no performance gains since the functions are synchronous executions (no timeouts or async tasks).

> hash

single hash is good

> how do I run it?

Only 1 mode, but the extent of the search can be adjusted internally: `searchLies` has an optional 2nd argument that can be omitted for max entropy.

```js
const response = getPrototypeLies(iframeWindow) // run
// ...ok to remove the iframe from the dom here
const { lieList, lieDetail, lieCount, propsSearched } =  response  // get properties from response

const hash = getHash(response) // example
```

Below is a modification to the internal search. The rendering contexts and element node APIs contain a lot of properties, so I targeted popular fingerprint properties in each, but for max entropy (every extension has suprises from time to time), we could remove the second argument (the object with target property) and everything will be searched. FF is pretty fast searching everything in ~200ms on my machine (chrome 2x slower). To see all the properties in a prototype, run `Object.getOwnPropertyNames(CanvasRenderingContext2D.prototype)` (for example).

> ~2400 properties 

A deep test page is a good option too. I plan to restore the creep test page to cover everything and more over time.

----

```js
    searchLies(() => AnalyserNode)
    searchLies(() => AudioBuffer)
    searchLies(() => BiquadFilterNode)
    searchLies(() => CanvasRenderingContext2D, {
        target: [
            'getImageData',
            'getLineDash',
            'isPointInPath',
            'isPointInStroke',
            'measureText',
            'quadraticCurveTo'
        ]
    })
    searchLies(() => Date)
    searchLies(() => Intl.DateTimeFormat)
    searchLies(() => Document, {
        target: [
            'createElement',
            'createElementNS',
            'getElementById',
            'getElementsByClassName',
            'getElementsByName',
            'getElementsByTagName',
            'getElementsByTagNameNS',
            'referrer',
            'write',
            'writeln'
        ]
    })
    searchLies(() => DOMRect)
    searchLies(() => DOMRectReadOnly)
    searchLies(() => Element, {
        target: [
            'append',
            'appendChild',
            'getBoundingClientRect',
            'getClientRects',
            'insertAdjacentElement',
            'insertAdjacentHTML',
            'insertAdjacentText',
            'insertBefore',
            'prepend',
            'replaceChild',
            'replaceWith',
            'setAttribute'
        ]
    })
    searchLies(() => Function, {
        target: [
            'toString',
        ]
    })
    searchLies(() => HTMLCanvasElement)
    searchLies(() => HTMLElement, {
        target: [
            'clientHeight',
            'clientWidth',
            'offsetHeight',
            'offsetWidth',
            'scrollHeight',
            'scrollWidth'
        ]
    })
    searchLies(() => HTMLIFrameElement)
    searchLies(() => IntersectionObserverEntry)
    searchLies(() => Math)
    searchLies(() => MediaDevices)
    searchLies(() => Navigator)
    searchLies(() => Node, {
        target: [
            'appendChild',
            'insertBefore',
            'replaceChild'
        ]
    })
    searchLies(() => OffscreenCanvasRenderingContext2D, {
        target: [
            'getImageData',
            'getLineDash',
            'isPointInPath',
            'isPointInStroke',
            'measureText',
            'quadraticCurveTo'
        ]
    })
    searchLies(() => Range)
    searchLies(() => Intl.RelativeTimeFormat)
    searchLies(() => Screen)
    searchLies(() => SVGRect)
    searchLies(() => TextMetrics)
    searchLies(() => WebGLRenderingContext, {
        target: [
            'bufferData',
            'getParameter',
            'readPixels'
        ]
    })
    searchLies(() => WebGL2RenderingContext, {
        target: [
            'bufferData',
            'getParameter',
            'readPixels'
        ]
    })
```

**Comment by Thorin-Oakenpants** [comment]
wooo .. slow down there young fella

- added - https://github.com/arkenfox/TZP/blob/master/js/prototypeLies.js
   - I used https://github.com/abrahamjuliot/web-fingerprinting/blob/febe65ffc65ed22797379003cce5979be392894e/prototypeLies.js which is 2 hours old
- note
   - cleaned up indents to spaces 🥇 
   - added `'use strict';`  👍 
   - wrapped in two commented out functions, but I guess we only need one function which returns a promise that it's finished?
- data currently just outputs to console (cuz it autoruns, my wrapper functions are commented out)
- it's the second file after fonts to be requested
- I can call the function whenever I like (e.g. when it tells me it's landed, or under my global section rerun code: which sounds like the best since that's when I start the overall timer
- > run it as a promise
   -  that's just so I know the function has finished and can then kickoff the rest of the tests

**Comment by Thorin-Oakenpants** [comment]
So I just need to engineer how to run it on demand

**Comment by Thorin-Oakenpants** [comment]
PS: btw, awesome work .. have some cake 🍰 beer 🍺 a medal 🥇 a hug 🤗 and a cat 🐈

**Comment by Thorin-Oakenpants** [comment]
ok, am onto this .. using a single function and returning a promise.. I think I can take this from here

**Comment by Thorin-Oakenpants** [comment]
wait a minute...

```
	const {
		lieList,
		lieDetail,
		lieCount,
		propsSearched
```

can I instead make these into global variables? So I can simply call them when clicking on `[x lies...]` for details?

**Comment by abrahamjuliot** [comment]
global variables will work

**Comment by Thorin-Oakenpants** [comment]
ok propsSearched is throughout the code, and I don't want to mess anything up (for future changes) : instead I will just store a copy in globals.js, and reset them as needed
- e.g. protoLies = lieList; protoCount = lieCount, protoProps = propsSearched .. etc

**Comment by Thorin-Oakenpants** [comment]
on tzp

chrome 88 reports no lies
opera 73 reports 9 lies
```
Document.createElement: Array(3)
0: "l: descriptor keys should only contain "name" and "length""
1: "m: own property names should only contain "name" and "length""
2: "n: own keys names should only contain "name" and "length""
length: 3
__proto__: Array(0)
Element.setAttribute: Array(3)
0: "l: descriptor keys should only contain "name" and "length""
1: "m: own property names should only contain "name" and "length""
2: "n: own keys names should only contain "name" and "length""
length: 3
__proto__: Array(0)
Node.appendChild: Array(3)
0: "l: descriptor keys should only contain "name" and "length""
1: "m: own property names should only contain "name" and "length""
2: "n: own keys names should only contain "name" and "length""
length: 3
```

**Comment by abrahamjuliot** [comment]
Opera adds the `call` property to the descriptors on at least 6 functions:
```
Document.createElement
Document.createTextNode
Document.querySelector
Element.querySelector
Element.setAttribute
Node.appendChild
```

**Comment by Thorin-Oakenpants** [comment]
While I was awaiting your thoughts on this, my gut reaction was to use the opera hash to detect isOpera :) and to then clear the lies and report none - but your idea is better, just leave em there !!

**Comment by Thorin-Oakenpants** [comment]
How come you got six and I only got three? Are you holding out on me? I see you did some more changes ... if anything applies here, just add it if you can

**Comment by abrahamjuliot** [comment]
6 are revealed if we search everything in `Document`, Element`, and `Node`. On slower machines, the additional performance time is significant on these beasts, so I recommend searching everything only on a test page. But, depending on what the test page reveals overtime, we could manually target more properties.

> changes

I'm testing a recent Cydec extension release and there appears to be an extremely unusual tampering behavior within the iframe prototype: it is stripped of nearly all native properties and fakes properties are added that should not be there. It's caught inderectly on TZP and interpreted as a failed execution tampering lie, but my change catches and handles it directly on the on the toString test by try catching `iframeWindow.Function.prototype.toString` which does not exist in Cydec. I presume this is an unintended, so I would only recommend this change as a target test on this extension. In any case, on TZP we catch all potential errors like this an render them as tampering lies due to failed execution.

However, I will include this if I add any significant changes that merge to the web-fingerprinting repo.

**Comment by abrahamjuliot** [comment]
I have a [new test](https://github.com/abrahamjuliot/web-fingerprinting/blob/940318b31a431059a1c4e50c8dbc715e71e79939/prototypeLies.js#L247) that catches attempts to modify `toString` including any attempts to mask `Function.prototype.toString` with a `Proxy`. You should see the lie count increase on Chameleon and a dramatic increase on Cydec. I will add this to TZP with the iframe `toString` try catch.

**Comment by Thorin-Oakenpants** [comment]
@abrahamjuliot 

so cydec has a FF extension (I didn't know that): that makes it easier for me to test

anyway: cydec lies about hardwareConcurrency and maxTouchPoints: these are not listed in prototype lies: can we add them

**Comment by abrahamjuliot** [comment]
✔ cb25184

This will catch:
```
enumerateDevices
getUserMedia
getDisplayMedia
vendorSub
productSub
vendor
maxTouchPoints
connection
plugins
hardwareConcurrency
appCodeName
appName
appVersion
platform
product
userAgent
sendBeacon
deviceMemory
buildID
activeVRDisplays
oscpu
```

**Comment by Thorin-Oakenpants** [comment]
OK, this sucks

```
// check lies later in any function
//lieList.includes('HTMLCanvasElement.toDataURL') // returns true or false
//lieDetail['HTMLCanvasElement.toDataURL'] // returns the list of lies
```

in a different js file
```js
   if (lieList.includes("Navigator.platform")) {uaBS = true}
   // ReferenceError: lieList is not defined
```

Do I need to stick prototype lies into generic.js?
```

**Comment by Thorin-Oakenpants** [comment]
seems like cydec is not reporting any `navigator` lies

**Comment by abrahamjuliot** [comment]
Hmmm... I entered `protoList.includes("Navigator.platform")` in the console and it catches Cydec in FF and Chrome.

**Comment by Thorin-Oakenpants** [comment]
sheeshus: just spent an eternity trying to work out why iframe checks weren't returning the real values

I threw `User-Agent Switcher and Manager` onto Chrome, so I could check if navigator prototype lies were showing up (they do) - see pic.

I spoofed as FF 86 on chrome, and all seven iframe methods returned the same hash: but not the real one. The difference is iframes only picked up that `oscpu` was fake: the other four are spoofed correctly

![partial-iframe](https://user-images.githubusercontent.com/16656956/110726679-bb198980-8211-11eb-9816-b754c06ca310.png)

So now I no longer use iframes to return the real values. Hindsight is wonderful - https://github.com/arkenfox/TZP/commit/5322d4693ed7c7ef64a1bf11b1cc62b873788241

---

Right .. so `User-Agent Switcher and Manager` shows five navigator lies. Cydec shows none. Maybe don't use the console to test as that is privileged : it is in firefox, probably the same in chrome ?

**Comment by Thorin-Oakenpants** [comment]
Pretty sure this chrome on windows is not opera borked edition or whatever on macos. But I can't detect it and return a single metric that it is a lie

https://github.com/arkenfox/TZP/blob/5322d4693ed7c7ef64a1bf11b1cc62b873788241/js/screen.js#L2799-L2817


![borked](https://user-images.githubusercontent.com/16656956/110727387-139d5680-8213-11eb-8aca-152dfaa737dc.png)

**Comment by Thorin-Oakenpants** [comment]
console test: and I definitely ain't using trident on an IE mobile
![console](https://user-images.githubusercontent.com/16656956/110728047-3da34880-8214-11eb-9e6c-7a7fe1dbe8d4.png)

**Comment by Thorin-Oakenpants** [comment]
installed a brand new chrome 64bit, added cydec, loaded TZP, same result: so it;s not my browser setup. Can you not replicate?

**Comment by abrahamjuliot** [comment]
That is strange. I'm inclined to say it's Cydec. Maybe 32-bit Chrome limits mad code. There could be a memory leak or something.

This is the "y" test. It should catch Cydec in the console. 

```js
console.log('lied:', 'value' in Object.getOwnPropertyDescriptor(Navigator.prototype, 'userAgent'))
```

**Comment by Thorin-Oakenpants** [comment]
> Maybe 32-bit Chrome limits mad code

I tried in an installed 64 bit, so that's not it. I'm going to take a break and come back to this later, and try that "y" test

**Comment by Thorin-Oakenpants** [comment]
tested creepy

9 errors
![errors](https://user-images.githubusercontent.com/16656956/110732071-59f6b380-821b-11eb-8a6c-c1b973be0de5.png)

this never finishes
![loading](https://user-images.githubusercontent.com/16656956/110732094-6713a280-821b-11eb-96fb-d0ca402a6ef3.png)

only 545 lies, only one navigator property
![545lies](https://user-images.githubusercontent.com/16656956/110732112-6da21a00-821b-11eb-9ffe-9724eb41d0fa.png)

WTF? Can you replicate?

**Comment by abrahamjuliot** [comment]
I'm not sure, but it could be Cydec is set to minimum spoofing mode. I tried fiddling around with the setting and minimum reduces the lies to ~550 (varies across versions). In the Profile tab- select `Standard`, and save (ignore the overrides message). The `Total` setting seems to yield the same effect as `Minimum`. At first, I could not figure out how to reset the setting back to standard, so I just deleted and reinstalled it, and the default setting is `Random` which oddly is the same as `Standard`. 

> 9 errors

Those are okay as long as they are counted in captured. Some are duplicates, so the total reduced. It's on my to-do to handle the version fetch test not finishing.

**Comment by Thorin-Oakenpants** [comment]
9 console errors: it's breaking creepJS = speech failed, audio failed ... I don't think I have two screwed up chromes: 1 up to date portable 32bit, 1 freshly installed vanilla 64bit. Creepy also fails to finish calculating the user agent stuff. It pretty much breaks creepy. Maybe you can't replicate what I'm seeing. It's **_definitely_** cydec, because none of those issues happen when it is disabled.

Maybe all that is extra noise and a separate problem?

---

It's weird that I cannot get navigator lies. Not a single one in TZP and only one on creepy in Chrome. This is nuts, because you can't replicate.

I think we should just focus on the actual prototypeLies, and try and debug what is going on. OK, just clicked - we can use
- https://abrahamjuliot.github.io/creepjs/tests/prototype.html

brb

**Comment by Thorin-Oakenpants** [comment]
^^ eeek, it needs some console outputs

**Comment by Thorin-Oakenpants** [comment]
chrome: with cydec
```
2546 lies detected in 1215 of 2109 properties

2109 searched: 3bc1029d
1215 corrupted: 0d5b215e
2546 lies: 923e9015
```

a single navigator property
![nada](https://user-images.githubusercontent.com/16656956/110737223-e194f000-8224-11eb-883e-b3cd56d5c758.png)

without cydec
```
0 lies detected in 2149 properties

2149 searched: 9149477c
```

is dropping 40 properties normal (looks suspiciously like all those nav properties)? In Firefox with cydec it goes up, from 2017 to 2030

**Comment by Thorin-Oakenpants** [comment]
> At first, I could not figure out how to reset the setting

Same. It's a bit weird. I actually had Firefox on `random` and have now flipped it to `total`. Same on chrome. However, standard/random includes user agent nav stuff: and I could see the spoofing in the results on the test page. So the setting wasn't it

---

Alright ... let's forget about the cause, and just add the fix.

This works: cydec lying it's ass off on chrome
![bingo](https://user-images.githubusercontent.com/16656956/110738429-2e79c600-8227-11eb-8d03-4a09bbb26f40.png)

So in prototypeLies.js, we sh/c/would want to use this `y` test to populate lieList, lieDetail etc - for just "navigator" properties. Can we add that in to harden the function against fuckery?

**Comment by abrahamjuliot** [comment]
That might fix it. I'll take a look. Let's craft it on the creep test page and see if it works. I should have something tomorrow night or Friday.

I need to test more on that `Total` mode. Happy to see the CSS Media test survived.

**Comment by Thorin-Oakenpants** [comment]
FWIW, this is my list: but I guess that was on standard/random. I also have one for brave. But ultimately, they will all be solved when I get my shit together and do the three bullet points below

Obviously, I'm aiming to do all of this to **_everything_** possible
- mathematical proofs = known lies
- iframe & worker tests = known lies
- all prototype lies

I'm a bit all over the place. I'm just going to go section by section. You on the other hand have a global multi-iframe, global workers (plural?) AFAICT. Like I said once before, if I started this today, I probably would have engineered it differently: but I do like being able to re-run a section

```
Cydec
=====

DONE
====

general
  - isFF hardening
feature detection
  - errors: bypassed with newFn
  - version interference looks solid
  - math: two pass
user agent
  - navigator: very solid uaBS for FF: but would not detect minor changes like Win7 to Win10
canvas
  - two pass and known values
domrect
  - two pass, shift, and known domrect values

TODO
====

user agent
  - navigator: awaiting prototype lies
feature detection
  - math: needs a better method, especially when we add the new v68+ test of 150+ items
devices
  - hardwareConcurrency
  - maxTouchPoints
headers
  - beacon
  - connection
screen
  - pixelDepth
  - screenDepth
languages
  - timezone
  - language
  - additional languages
  - locales
  - date/time
audio
  - audio
  - audio2?
```

That was just things that changed between tests on TZP

**Comment by Thorin-Oakenpants** [comment]
FYI: here's my brave checklist

```
Brave
=====

https://dev-pages.brave.software/farbling.html

DONE
===
canvas (noise detected)
	- 97f1593eb4ebb43744a8c98c682a4025e0140d39
plugins (creepy pluginLies)
	- fake random
hardwareConcurrency (if pluginBS then this is also fake)
	- 2 to 8 random
UA for now (if isBrave sanitize spaces)
	- uses a random number of trailing spaces on two properties
	- they may add inserting these into the middle of the string

TODO
====

webgl
	- is randomized
media devices
  - to check
audio (random)
	- autorun = random
	- manual test: keys = not random
	- manual test: numbers = random
textmetrics
  -integers on strict I think and all the same
		- 10: "tm_actualBoundingBoxAscent:46d1548a101ef8e7bb50fab11917c53096c1ac63"
		- 11: "tm_actualBoundingBoxDescent:46d1548a101ef8e7bb50fab11917c53096c1ac63"
		- 12: "tm_actualBoundingBoxLeft:46d1548a101ef8e7bb50fab11917c53096c1ac63"
		- 13: "tm_actualBoundingBoxRight:46d1548a101ef8e7bb50fab11917c53096c1ac63"
		- 17: "tm_fontBoundingBoxAscent:46d1548a101ef8e7bb50fab11917c53096c1ac63"
		- 18: "tm_fontBoundingBoxDescent:46d1548a101ef8e7bb50fab11917c53096c1ac63"
		- 21: "tm_width:46d1548a101ef8e7bb50fab11917c53096c1ac63"

WATCH
=====

mime types
	- not random yet
domrect
	- not random yet
```

**Comment by abrahamjuliot** [comment]
Okay, I'm able to replicate this in `Total` spoofing mode. 🤦‍♂️🥴

![2021-03-11 18_29_08-DimScreen Screen](https://user-images.githubusercontent.com/6946045/110883170-ea72d980-8297-11eb-89ba-b319b723e349.png)

Indeed, the issue is Cydec. `Object.getOwnPropertyNames(Navigator.prototype)
` should return the same as `Object.keys(Navigator.prototype)`.

![image](https://user-images.githubusercontent.com/6946045/110886541-ada9e100-829d-11eb-837a-076194fd59e8.png)

To fix this, I now use both and reduce them to unique names.

```js
;[...new Set([
    ...Object.getOwnPropertyNames(interfaceObject),
    ...Object.keys(interfaceObject)
])]
```
Try the test. If it works on your end, I'll apply a patch.

**Comment by Thorin-Oakenpants** [comment]
^^ worksforme 👍 just tried `total` mode cydec and it lists all the nav properties

```
was: 2546 lies detected in 1215 of 2109 properties  <-- from about sic posts up
now: 2695 lies detected in 1274 of 2161 properties
```

1274 - 1215 = 59 properties

60 nav items listed: (previous test had a single nav property) - so there's the diff
```
2 Navigator.doNotTrack
1 Navigator.vendorSub
1 Navigator.productSub
1 Navigator.vendor
1 Navigator.maxTouchPoints
2 Navigator.userActivation
2 Navigator.geolocation
1 Navigator.connection
2 Navigator.mimeTypes
2 Navigator.webkitTemporaryStorage
2 Navigator.webkitPersistentStorage
1 Navigator.hardwareConcurrency
2 Navigator.cookieEnabled
1 Navigator.appCodeName
1 Navigator.appName
1 Navigator.appVersion
1 Navigator.platform
1 Navigator.product
1 Navigator.userAgent
1 Navigator.language
2 Navigator.onLine
2 Navigator.webdriver
8 Navigator.getBattery
9 Navigator.getGamepads
2 Navigator.javaEnabled
2 Navigator.vibrate
2 Navigator.serial
2 Navigator.scheduling
2 Navigator.mediaCapabilities
2 Navigator.permissions
2 Navigator.locks
2 Navigator.wakeLock
2 Navigator.usb
2 Navigator.mediaSession
2 Navigator.clipboard
2 Navigator.credentials
2 Navigator.keyboard
2 Navigator.mediaDevices
2 Navigator.storage
9 Navigator.serviceWorker
1 Navigator.deviceMemory
2 Navigator.hid
2 Navigator.presentation
2 Navigator.userAgentData
2 Navigator.xr
2 Navigator.canShare
2 Navigator.share
3 Navigator.registerProtocolHandler
3 Navigator.unregisterProtocolHandler
2 Navigator.getInstalledRelatedApps
2 Navigator.clearAppBadge
2 Navigator.setAppBadge
3 Navigator.getUserMedia
2 Navigator.requestMIDIAccess
3 Navigator.requestMediaKeySystemAccess
3 Navigator.webkitGetUserMedia
9 Navigator.SharedWorker
9 Navigator.Worker
9 Navigator.getVRDisplays
1 Navigator.activeVRDisplays
```

**Comment by Thorin-Oakenpants** [comment]
hmmm .. those don't look sorted

**Comment by abrahamjuliot** [comment]
✔ 13cd0e4 (handle Cydec)
✔ 741f8b7 (sort) - good point. I had not considered this.

**Comment by Thorin-Oakenpants** [comment]
updated my cydec list above and added some total mode fuckery: sheesh it's a bit of a PITA to try and change between modes

not sure if this is all of it: didn't dig into screen or language section diffs
```
feature detection: TOTAL MODE
	- errors: error 2:  newFn(`window.onload = ${foobar}()`)
			- expected (fix  true): TypeError: can't access property "value", foo is null
			- expected (fix false): TypeError: foo is null
			- got: SyntaxError: expected expression, got ')'
fonts: TOTAL MODE
	- unicode glyphs: all heights are 2300, widths are random
	- domrect
	- textmetrics (width only)
css: TOTAL MODE
	- computed styles: changed from 1063 to 1064
	- prefers reduced-motion/color-scheme = blocked
		- we will add a css rule element and read back the value
misc: TOTAL MODE
	- iframe properties: went from 833 to 836 items
	- navigator properties (nightly FF)
		- removed `plugins`, `languages`
		- added `SharedWorker`, `Worker`, `getBattery`, `connection`
```

So the first one that jumps out is the misc: nav properties: how do we harden this?

https://github.com/arkenfox/TZP/blob/fdc65a63ea36cd7dc657f7a0c28428645507f7bc/js/misc.js#L74-L88

Then computed styles change needs a debug, as well as iframe properties. Everything else I think I can handle: e.g. I might use a different test for error 2, textmetrics width I have a mathematical PoC (don't see a prototype lie for this?), css matchMedia fuckery I will add a getComputedStyle for an element as a third check

**Comment by abrahamjuliot** [comment]
As far as I know, the iframe window, navigator, and computed style properties are stable per execution. but the user can turn off the settings that alter these.

> misc nav properties

Here are some unusual lies:
- `getBattery` should not appear in FF
- `activeVRDisplays` should not appear in Chrome
- `('getBattery' in navigator && 'activeVRDisplays' in navigator)` should not be true
- `SharedWorker` and `Worker` should never exist in `navigator`

> window + iframe window

- `serviceWorker` should never exist in `window` or `contentWindow`
- `mozRTCPeerConnection` should not appear in Chrome
- Cydec also uses public variables:
  - `hdcd_canvas_getctx` is in `window` and `contentWindow`
  - more `hdcd...` variables appear in `window`
 
![Screenshot 2021-03-13 at 4 16 30 PM](https://user-images.githubusercontent.com/6946045/111053305-a27ac080-8417-11eb-9a4e-708675f2edb2.png)

> TextMetrics

`protoDetail['TextMetrics.width']` should return lies (tested in `Standard` and `Total`).

![image](https://user-images.githubusercontent.com/6946045/111054662-1a9ab380-8423-11eb-9a1b-612f0cc1778a.png)

**Comment by Thorin-Oakenpants** [comment]
> but the user can turn off the settings that alter these

sometimes: depends on the property. Not concerned with that: e.g. connection is behind a pref in FF and has a different state for desktop vs mobile.

Because we know the engine, there are some obvious lies there (e.g battery in FF), but I was assuming we could detect lies here programmatically (vs relying on a slip up in the extension), and discard the BS. i.e I was hoping that function could be hardened to always return the truth

> textmetrics

Yeah, saw that afterwards. The interesting thing is that even though the other textmetric properties are in prototype lies, only width is being used. This is why I am more keen on relying solely on unbeateable mathematical proofs where possible. A prototype lie only means we can't trust it, not that it is actually a lie

> the iframe window, navigator, and computed style properties are stable per execution

clearly not. in my test, the only change was enabling/disabling cydec in Total mode

**Comment by abrahamjuliot** [comment]
> returning nav properties

I use 
```js
const keys = Object.keys(Object.getPrototypeOf(navigator))
```

> hardened to always return the truth

The only solution I'm aware of is to determine if there is an unprotected iframe and then get the true set of properties there. 

However, if we know the last descriptor key, we can determine what properties were added. If properties are added (via a script), they are appended to the order of the descriptor object keys. 

This example works in FF `Navigator` since we know the last key is `constructor`. 
```js
keys = Object.keys(Object.getOwnPropertyDescriptors(Navigator.prototype)) // reported descriptor keys
lastKey = keys.indexOf('constructor') // the last descriptor key
fakeKeys = keys.slice(lastKey+1) // the fake set of keys
trueKeys = keys.slice(0, lastKey+1) // the true set of keys
```
Another bonus of using `getOwnPropertyDescriptors` is the order of the true keys is not affected by the fake keys. But, I'm not sure how stable this is across versions. From what I observed in FF, `constructor` is consistently the last descriptor key on a given interface. The last key on `Navigator` in Chrome is `webkitGetUserMedia` which is likely not stable.

**Comment by Thorin-Oakenpants** [comment]
iframe properties might have been because I had the console open for the second test: it went from 833 to 836 : although I thought we had already fixed that: anyway, I will make the counts (for lots of things) clickable to log the data to console

nav, looks exciting: will give it a whirl .. but we still need to handle blink (no idea about webkit)

**Comment by Thorin-Oakenpants** [comment]
sweeet
```js
function get_nav_prototype() {
	let hash, keys, keyWord = "", lastKeyIndex
	try {
		keys = Object.keys(Object.getOwnPropertyDescriptors(Navigator.prototype))
		if (isFF) {
			// FF: constructor is always last
			keyWord = "constructor"
		} else if (isEngine == "blink") {
			// chromium: not sure if this is always last
			keyWord = "webkitGetUserMedia"
		}
		if (keyWord == "") {
			// everything else: who knows
			lastKeyIndex = keys.length
		} else {
			lastKeyIndex = keys.indexOf(keyWord) // the index of the last descriptor key
		}
		let fakeKeys = keys.slice(lastKeyIndex+1) // the fake set of keys
		let trueKeys = keys.slice(0, lastKeyIndex+1) // the true set of keys
		hash = sha1(trueKeys.join())
		dom.nProto.innerHTML = hash + s18 +"["+ trueKeys.length +"]"+ sc
		dom.nProto2 = trueKeys.join(", ")
		dom.nProto2.style.color = zshow
		// temp debug for a while
		if (fakeKeys.length > 0) {console.log("fakeKeys", fakeKeys)}
	} catch(e) {
		hash = zB0
		dom.nProto = hash
		dom.nProto2 = ""
	}
	return "navigator_properties:" + hash
}
```


This is now working sweet as in FF. Chromium works, assuming `webkitGetUserMedia` is always the last real item. 

I noticed FF spoofing seems to be consistent (we no longer need to worry about dropped items)
- it always adds: `SharedWorker`, `Worker`, `getBattery`, `connection`
- maybe it's slightly variable

but in chromium it depends on what it is spoofing the browser as
- I think it always adds: `SharedWorker`, `Worker`, `getVRDisplays`, `activeVRDisplays`
- seen it also add `buildID` and `oscpu` (presumably when it was claiming to be gecko)

How certain are we that chromium always reports `webkitGetUserMedia` last ? As for webkit, I'm not worried about extensions in webkit, unless you are?

**Comment by Thorin-Oakenpants** [comment]
OK, counts are clickable: they output to console.log: I'll expand this to more items

here's cydec
![clickclick](https://user-images.githubusercontent.com/16656956/111125652-c0583a80-8569-11eb-9c44-df5dc85adf6c.png)

**Comment by Thorin-Oakenpants** [comment]
cydec `total` added the following in both FF (tested nightly) and on chrome to `iframe.contentWindow` properties

it went from 833 to 836 items in my test
```
hdcd_canvas_getctx
serviceWorker
webkitRTCPeerConnection
```

That was just one test each, although I have done the FF one lots of times and it was always 836. But now I can click and copy for compare, I can easily get the diff

**Comment by abrahamjuliot** [comment]
> webkit

Same. I only check WebKit to make sure nothing breaks. 

> webkitGetUserMedia

`webkitGetUserMedia` is inconsistent. For example, `wakeLock` is the last index in Chrome 84.

We could target Cydec directly in Chrome.
```js
knownPoison = ["SharedWorker", "Worker", "buildID", "getVRDisplays", "activeVRDisplays", "oscpu"]
keys = Object.keys(Object.getOwnPropertyDescriptors(Navigator.prototype))
backEndKeys = keys.slice(-knownPoison.length)
trustedKeys = backEndKeys.filter(x => !knownPoison.includes(x)) // remove known poison keys
frontEndKeys = keys.slice(0, -knownPoison.length)
restoredKeys = [...frontEndKeys, ...trustedKeys] // the pure list of keys
```

This works in FF too. Just add `getBattery` and `connection` to `knownPoison` (remove `oscpu` and `buildID`). This should be stable as long as `getVRDisplays` and  `activeVRDisplays` do not belong at the back end test keys, and ideally, `knownPoison` should remain short in length. But, the `constructor` test above should catch-all in FF.

**Comment by Thorin-Oakenpants** [comment]
alrighty
```js
	let keys = Object.keys(Object.getOwnPropertyDescriptors(Navigator.prototype))
	let trueKeys = keys
	let lastKeyIndex = keys.length
	let fakeKeys = [], fakeStr = ""

	if (isFF) {
		// FF: constructor is always last
		lastKeyIndex = keys.indexOf("constructor")
		trueKeys = keys.slice(0, lastKeyIndex+1)
		fakeKeys = keys.slice(lastKeyIndex+1)
	} else if (isEngine == "blink") {
		// chromium last key is inconsistent
		let knownPoison = ["SharedWorker", "Worker", "buildID", "getVRDisplays", "activeVRDisplays", "oscpu"]
		// split knownPoison from trueKeys into fakeKeys
		trueKeys = keys.filter(x => !knownPoison.includes(x))
		fakeKeys = keys.filter(x => knownPoison.includes(x))
	}
```

Not foolproof: the knownPoison list needs to catch all BS: if one gets through (e.g. randomized gibberish). Well, at least FF is solid

**Comment by Thorin-Oakenpants** [comment]
sendBeacon isn't listed

FF control
- cydec off
- make sure `beacon.enabled` = true
- load TZP
- header section>details - beacon = enabled
- misc section> nav objects includes `sendBeacon`

TEST
- cycec on (mine was in total mode but pretty sure this is also tampered with in standard)
- make sure `beacon.enabled` = true
- load TZP
- header section>details - beacon = disabled
   - console: `protoList.includes("Navigator.sendBeacon")` = `false`
- misc section> nav objects includes `sendBeacon`

**Comment by Thorin-Oakenpants** [comment]
not sure what I'm doing here. Yes, this is duplicated data in nav proporties, but I want this info up in the header/networking section as well. Besides, I can use it for other checks elsewhere

generic function: can we make this cleaner?
```js
function check_navObject(property) {
	let answer = ""
	try {
		answer = (`lied:`, `value` in Object.getOwnPropertyDescriptor(Navigator.prototype, property))
		answer = true
	} catch(e) {
		// TypeError right-hand side of 'in' should be an object, got undefined
		//console.debug("navObject error", e.name, e.message)
		answer = false
	}
	return answer
}
```

usage?
```js
//WAS THIS
	//let beacon = ""
	//try {beacon = (navigator.sendBeacon ? zE : zD)} catch(e) {beacon = zB0}

// NOW THIS
	let beacon = (check_navObject("sendBeacon") ? "enabled" : "disabled")
```

So this is just returning if the API is enabled, right?
- In a binary result like above, it replaces checking the actual value (note: I am not testing if a beacon is actually sent)
- In an API where the value is more than binary, it would simply replace e.g. `if ("doNotTrack" in navigator)`, `if ("connection" in navigator)`
   - I have a lot of these sorts of checks, so anything to harden it would be good

Right?

**Comment by Thorin-Oakenpants** [comment]
alrighty, replaced 27 instances in `"property" in navigator` with `check_navObject("property")`

**Comment by abrahamjuliot** [comment]
> protoLies does not contain sendBeacon

I'll see if I can create a patch for this. I also miss `languages` and `plugins`. I believe all I need to do is change `Object.keys(Navigator.prototype)` to `Object.keys(Object.getOwnPropertyDescriptors(Navigator.prototype))` in the test.

> generic function: can we make this cleaner?

```js
function check_navObject(property) {
	return !!Object.getOwnPropertyDescriptor(Navigator.prototype, property)
}
```

**Comment by Thorin-Oakenpants** [comment]
FWIW: you can click shit on TZP now
![clicky](https://user-images.githubusercontent.com/16656956/111565180-dc8af000-8792-11eb-8d76-7e61de8758d0.png)

**Comment by Thorin-Oakenpants** [comment]
```js
function check_navObject(property) {
	return !!Object.getOwnPropertyDescriptor(Navigator.prototype, property)
}
```

fails for at least beacon with cydec
- cydec off, `beacon.enabled` = true = returns true = ✅ 
- cydec off, `beacon.enabled` = false = returns false = ✅
- cydec on, `beacon.enabled` = true = returns true = ✅ 
- cydec on, `beacon.enabled` = false = returns true = ❌

**Comment by Thorin-Oakenpants** [comment]
https://arkenfox.github.io/TZP/tests/navigator.html .. I'm confused

**Comment by Thorin-Oakenpants** [comment]
@abrahamjuliot 

I was going through my code trying to remove instances of multiple spaces, and I noticed this one: is that OK to change to a single space?

https://github.com/arkenfox/TZP/blob/0dca3ec7fdd57fd76b709ebd7398d2d962bc4490/js/prototypeLies.js#L280-L281

**Comment by abrahamjuliot** [comment]
Yes, that spacing needs a clean-up. Also, these identifier strings can be reduced and customized. For example, instead `a: ...` and `b: ...` you can change this to `apples` and `bananas`, and then comment/reference the word in the code test section for detail. My attempt to describe the test in one line is at best only a hint to what the test is doing.

`["jackfruit"]: getDescriptorLie(apiFunction)`

**Comment by Thorin-Oakenpants** [comment]
Cydec total mode in all three computed styles adds a `setProperty` at the end in FF

![styles](https://user-images.githubusercontent.com/16656956/113381464-91660500-936e-11eb-87f6-7f826b4d768d.png)

So... this looks like one of those isFF, anything after constructor is poison? @abrahamjuliot 

I have yet to check what it does on chromium

**Comment by abrahamjuliot** [comment]
Yes, from what I have observed, that is correct.

**Comment by Thorin-Oakenpants** [comment]
chromium normal
```
 getComputedStyle | aed8bcd135c6dd510e3f9773007774b677ede601 [1049|0|308]
HTMLElement.style | e89099a750e9d5d2dc4b6a5e5d4c4d54e252f00a [1049|0|308]
CSSRuleList.style | e89099a750e9d5d2dc4b6a5e5d4c4d54e252f00a [1049|0|308]
```
chromium cydec
```
 getComputedStyle | 375468c750f5da780cb83dce4722e080420c71c6 [1049|0|308]
HTMLElement.style | 34192a4558d35541fe95833e1011eb7b5cedbacc [1049|0|308]
CSSRuleList.style | 34192a4558d35541fe95833e1011eb7b5cedbacc [1049|0|308]
```

I think these are the same results, just a different order

**Comment by abrahamjuliot** [comment]
I use `JSON.stringify` to log and diff check objects in Chrome.

Cydec may attempt to poison this. I'll keep a close watch on it, but I think we can target Cydec if there is a pattern.

**Comment by Thorin-Oakenpants** [comment]
Yeah. From my perspective, the more "real values" I can expose, the less noise in the data sets. Same with randomness, or "trash" (untrusted) - i.e I might not be able to expose the true values, but I can render it to a static value.

In this case, cydec total mode results are already very different so it probably makes little to no difference. But should it add random noise then it becomes a slight problem: on the backend you can manipulate the data to exclude this noise

**Comment by Thorin-Oakenpants** [comment]
@abrahamjuliot does [bugzilla 1701904](https://bugzilla.mozilla.org/show_bug.cgi?id=1701904) and alluded other changes affect anything

**Comment by abrahamjuliot** [comment]
I checked Nightly latest and this does not appear the affect anything.

**Comment by Thorin-Oakenpants** [comment]
Hmmm .. I wonder if it's worth asking @evilpie about what this all entails: but you'll have to do the talking

**Comment by Thorin-Oakenpants** [comment]
Oh, OT: I saw Creepy listed on Cydec .. no-one ever bothers to mention me or TZP *sigh*

**Comment by Thorin-Oakenpants** [comment]
> I checked Nightly latest and this does not appear to affect anything.

Yeah, I looked at a bunch of other tickets to do with all this, and I think it's just refactoring the something something - like part of Warp and Stencil?

---

@abrahamjuliot I think we're going to need a bigger or newer BS thread.

So I just noticed Cydec is now fucking with perf: I hardened isRFP for firefox, but I'm sure there's a simple way to detect the BS, e.g. a setTimeout. What do you think

Anyway: here's what cydec tries to fubar
- I use isRFP to detect it's fake
- cydec's returns are actually `0,0,0,0` and `100 ms` (like RFP)
- I still return "not zero" in the section metrics so it's stable
- I add a global lie for perf.mark
- perf.now is not used in the FP
![fake](https://user-images.githubusercontent.com/16656956/114315294-dc5de600-9aed-11eb-967d-4ccec307eac1.png)

Here's my hardened isRFP: not 100% perfect
- chk1+2: a user may indeed be using light and no-preference (they are default)
   - Cydec blocks these values in matchMedia: in other words Cydec can't spoof them (unless it injects css? IDK)
- chk3: a user may indeed have flipped `dom.enable_performance_navigation_timing`
```js
const get_isRFP = () => new Promise(resolve => {
	isRFP = false
	// skip
	if (!isFF) {return resolve("")}
	try {
		performance.mark("a")
		let r = performance.getEntriesByName("a","mark").length
			+ performance.getEntries().length
			+ performance.getEntries({name:"a",entryType:"mark"}).length
			+ performance.getEntriesByName("a","mark").length
			performance.clearMarks()
		isRFP = (r == 0) // <----------------------- cydec passes this: 0,0,0,0
		// extra checks
		if (isVer > 62) {
			let chk1 = getContent('#cssPRM',':after')
			if (chk1 !== "no-preference") {isRFP = false}
		}
		if (isVer > 66) {
			let chk2 = getContent('#cssPCS',':after')
			if (chk2 !== "light") {isRFP = false}
		}
		if (isVer > 77) {
			let chk3 = zD
			try {if (window.PerformanceNavigationTiming) {chk3 = zE}} catch(e) {}
			if (chk3 !== zD) {isRFP = false}
		}
		return resolve("")
	} catch(e) {
		console.error("get_isRFP", e.name, e.message)
		return resolve("")
	}
})
```

What are your thoughts? Class, discuss!

---

Also [cydec](https://www.cydecplatform.com/antifp.html) .. scroll down to `Resources` - CreepJS is listed (I guess TZP isn't worthy)

**Comment by abrahamjuliot** [comment]
This tampering can be caught with math.

```js
stablePerformance  = Math.trunc(performance.now() - performance.now()) === 0
```

or

```js
isCydec = (performance.now() - performance.now()) === -100
```

**Comment by Thorin-Oakenpants** [comment]
that's good for hardening isRFP and detecting `.now()` fuckery (on chrome). Not that I am care too deeply about non FF, but `.mark` is probably too hard to catch (and that's the one I use in the FP)

**Comment by Thorin-Oakenpants** [comment]
cydec fuckery with window properties (FF only)
![dodgy-as-fuck](https://user-images.githubusercontent.com/16656956/114499739-44a7e700-9c16-11eb-8fe0-b9995afb595a.png)

Not entirely sure how much of a game of whack-a-mole this might end up. I'm using an "allowlist" rather than a "blacklist" - time will tell if this is the best move - here's the initial commit: https://github.com/arkenfox/TZP/commit/167b0f74196b31936fd862c85878270e64565bbf

So
- AFAICT Firefox always ends with `Performance` (but `Event` can come after that depending on if the console is open: also in FF60 but not FF68 didn't check which version it changed `StyleSheetList`)
- So: get `Performance` position, anything after that, excluding `Event` and `StyleSheetList` is suspect
- Remove known legit items from the suspect list to discover the fake list

What I find interesting is the list of suspect items: it's like a **_cydec fuckery cheat sheet_**: I wonder if other extensions create suspect lists (and my allowlist needs additional items)

What are your thoughts on this approach?

**Comment by Thorin-Oakenpants** [comment]
hah: chameleon (but I didn't turn on all protections)
```
MISC: iframe properties suspect: 1c7ac6919d4e2741a80541dc881811d7c969912d 
Array(9) [ "Bmr7ol9h", "Intl", "Navigator", "MediaDevices", "Element", "AbstractRange", "Range", "Date", "chrome" ]
​```

chrome and Bmr7ol9h are bogus

**Comment by Thorin-Oakenpants** [comment]
Actually .. I wonder if it's possible to use build some known legit exceptions from prototype lies: here's Cydec: 

![knownGood](https://user-images.githubusercontent.com/16656956/114503913-8d16d300-9c1d-11eb-90e4-7e3c337954aa.png)

Hmmm, cydec's prototype lies using the first part up to the `.`
```
these are in cydec suspect
  "AnalyserNode",
  "AudioBuffer",
  "CanvasRenderingContext2D",
  "Date",
  "Element",
  "HTMLCanvasElement",
  "HTMLElement",
  "MediaDevices",
  "Navigator",

these should not exist in window props
  "DateTimeFormat",
  "RelativeTimeFormat",

these exist in window props
  "BiquadFilterNode",
  "DOMRect",
  "DOMRectReadOnly",
  "Document",
  "Function",
  "HTMLIFrameElement",
  "IntersectionObserverEntry",
  "Math",
  "Node",
  "Range",
  "SVGRect",
  "Screen",
  "TextMetrics",
  "WebGL2RenderingContext",
  "WebGLRenderingContext",
```

So of the 27 suspect properties, prototype lies would only eliminate 9 of them, and we'd have to be careful to not allow any incorrect items in like `DateTimeFormat`

A [full test](https://abrahamjuliot.github.io/creepjs/tests/prototype.html) at first glance seems to capture them all
